### ISSUE TITLE
Checkpointing broken in master

### ISSUE THREAD

**Issue by navjotk** [devitocodes/devito#1399]
[This](https://github.com/devitocodes/devito/blob/master/tests/test_gradient.py#L32) test is supposed to compare the checkpointed and non-checkpointed implementations to ensure they produce the same result. However, it is running in double precision and only checking up to `1e-5` ([default](https://numpy.org/doc/stable/reference/generated/numpy.allclose.html) `rtol` of `np.allclose()`)

It is, obviously, meaningless to run in double precision and only ensure correctness up to 5 significant figures. The correctness of Checkpointing pretty much relies on this test passing, and if the test is broken, the feature is most likely broken too. 

Detailed investigation ongoing.

**Comment by navjotk** [comment]
Update. Dynamic scheduling seems to be the culprit here. I'll push an updated test shortly that disables dynamic scheduling for the test, and makes the numerical comparison tighter.

**Comment by navjotk** [comment]
Another potential culprit: fast math

**Comment by navjotk** [comment]
During #1412 we realised that the test only passes with tighter bounds with `opt=('advanced', {'par-dynamic-work :1000})` does not work. Only works with `opt=('noop', {'par-dynamic-work :1000})`. Could be a bug in blocking or one of the other advanced passes. More investigation required.

**Comment by ggorman** [comment]
The generated code for for the gradient where the save case generates the code:
```C
        grad[x + 1][y + 1] += -(-2.0F*u[t0][x + 4][y + 4] + u[t1][x + 4][y + 4] + u[t2][x + 4][y + 4])*v[t0][x + 4][y + 4]/(dt*dt);
```
and the checkpointing code generates the code
```C
        grad[x + 1][y + 1] += -(-2.0F*u[time][x + 4][y + 4] + u[time - 1][x + 4][y + 4] + u[time + 1][x + 4][y + 4])*v[t0][x + 4][y + 4]/(dt*dt);
```

The source of the roundoff error that causes the test to fail is that the terms `u[t1][x + 4][y + 4] + u[t2][x + 4][y + 4]` and `u[time - 1][x + 4][y + 4] + u[time + 1][x + 4][y + 4]` are actually swapped relative to each other; ie the additions are carried out in a different order in both cases. By flipping the order of these two terms in the second case (ie `u[time + 1][x + 4][y + 4] + u[time - 1][x + 4][y + 4]`) the two gradients become bitwise reproducible.

**Comment by FabioLuporini** [comment]
addressed by #1438
